Sources/MongoSwift/ReadPreference.swift

```diff
@@ -0,0 +1,167 @@
+import libmongoc
+
+/// A class to represent a MongoDB read preference.
+final public class ReadPreference {
+
+    /// An enumeration of possible ReadPreference modes.
+    public enum Mode: String {
+        /// See https://docs.mongodb.com/manual/reference/read-preference/
+        case primary
+        case primaryPreferred
+        case secondary
+        case secondaryPreferred
+        case nearest
+
+        internal var readMode: mongoc_read_mode_t {
+            switch self {
+            case .primary:
+                return MONGOC_READ_PRIMARY
+            case .primaryPreferred:
+                return MONGOC_READ_PRIMARY_PREFERRED
+            case .secondary:
+                return MONGOC_READ_SECONDARY
+            case .secondaryPreferred:
+                return MONGOC_READ_SECONDARY_PREFERRED
+            case .nearest:
+                return MONGOC_READ_NEAREST
+            }
+        }
+
+        internal init? (readMode: mongoc_read_mode_t) {
+            switch readMode {
+            case MONGOC_READ_PRIMARY:
+                self = .primary
+            case MONGOC_READ_PRIMARY_PREFERRED:
+                self = .primaryPreferred
+            case MONGOC_READ_SECONDARY:
+                self = .secondary
+            case MONGOC_READ_SECONDARY_PREFERRED:
+                self = .secondaryPreferred
+            case MONGOC_READ_NEAREST:
+                self = .nearest
+            default:
+                return nil
+            }
+        }
+    }
+
+    /// A pointer to a `mongoc_read_prefs_t`.
+    internal var _readPreference: OpaquePointer?
+
+    /// The mode of this `ReadPreference`
+    public var mode: Mode {
+        let readMode = mongoc_read_prefs_get_mode(self._readPreference)
+
+        guard let mode = Mode(readMode: readMode) else {
+            preconditionFailure("Unexpected read preference mode: \(readMode)")
+        }
+
+        return mode
+    }
+
+    /// The tags of this `ReadPreference`
+    public var tagSets: [Document] {
+        guard let bson = mongoc_read_prefs_get_tags(self._readPreference) else {
+            return []
+        }
+
+        let wrapped = Document(fromPointer: bson)
+
+        return wrapped.values as! [Document]
+    }
+
+    /// The maxStalenessSeconds of this `ReadPreference`
+    public var maxStalenessSeconds: Int64? {
+        let maxStalenessSeconds = mongoc_read_prefs_get_max_staleness_seconds(self._readPreference)
+
+        if maxStalenessSeconds == MONGOC_NO_MAX_STALENESS {
+            return nil
+        }
+
+        return maxStalenessSeconds
+    }
+
+    /**
+     * Initializes a `ReadPreference` from a `Mode`.
+     *
+     * - Parameters:
+     *   - mode: a `Mode`
+     *
+     * - Returns: a new `ReadPreference`
+     */
+    public init(_ mode: Mode) {
+        self._readPreference = mongoc_read_prefs_new(mode.readMode)
```

Comment:
Since initialization from a mode enum will never throw, I think this constructor is helpful.

I did consider making this a `convenience` for the long-form constructor. I believe that would just require changing this to:

```swift
try! self.init(mode)
```

I originally expected I might need to pass `nil` as one of the two optional arguments (`tagSets` or `maxStalenessSeconds`), but Swift seemed to infer that I must be calling the other initializer instead of making a recursive call to this initializer. I'm not sure if that's due to `convenience` or some other clever delegation logic.

Reply:
On second thought, Swift actually didn't like `try! self.init(mode)`. I'm not sure what I was looking at when I posted the last comment.

This will remain as-is, as passing `nil` named arguments just to coerce selection of the other initializer seems a hackish.

Reply:
yes, since this is a one-liner I think it's fine to just have it as a separate/non-`convenience` init.